### PR TITLE
[IIO] setting caps and data configs

### PR DESCRIPTION
**Pull request by kparichay**
1. Finding caps from data configs based on the device channels
2. Fixating the caps for the pad
3. Filling _set and _get for caps
4. Added feature to merge channels data into 1 tensor

Signed-off-by: Parichay Kapoor <pk.kapoor@samsung.com>

### CODE REVIEW

File: gst/nnstreamer/tensor_source/tensor_src_iio.c [hunk 1]
```diff
@@ -870,6 +956,143 @@ gst_tensor_set_all_channels (GstTensorSrcIIO * self, const gint contents)
   return ret;
 }
 
+/**
+ * @brief find type based on bits used
+ * @param[in] channel_prop Channel property containing bits used
+ * @returns type of the data inferred using bits used by the channel data
+ */
+static tensor_type
+gst_tensor_src_iio_get_type_for_channel (GstTensorSrcIIOChannelProperties *
```
Comment (leemgs) [inline gst/nnstreamer/tensor_source/tensor_src_iio.c]: It's reasonable to me at least.

**Comment by kparichay** [comment]
> This _single_ patch touches two modules : **Common** and **IIO**.
> IMO, separating two patches looks much better then single one.
> Except that, I gave you +2.

I made another PR for the changes in **Common** module, remove the corresponding changes from here.
Once, it is merged, this PR will also pass all the checks.
I made another PR #1238 for the changes in **Common** module.
Corresponding changes of the **Common** module have been removed from here.
Once, #1238 is merged, this PR will pass the checks.

**Comment by kparichay** [comment]
Waiting for PR #1238 .

**Comment by myungjoo** [comment]
PTAL @leemgs @again4you

File: gst/nnstreamer/tensor_source/tensor_src_iio.c [hunk 2]
```diff
@@ -322,8 +339,69 @@ gst_tensor_src_iio_init (GstTensorSrcIIO * self)
...
+ * @returns >=0 number of valid entries in the info after merge
+ *         -1 failed due to missing extra dimension/mismatch shape/type
+ * @todo merge tensors (in stable order) into buckets based on type
+ */
+static gint
+gst_tensor_src_merge_tensor_by_type (GstTensorInfo * info, guint size)
+{
+  gint info_idx, base_idx, dim_idx;
+  gboolean mismatch = FALSE, dim_avail = TRUE;
+  gint ret = 0;
```
Comment (jaeyun-jung) [inline gst/nnstreamer/tensor_source/tensor_src_iio.c]: I think ret is unnecessary in this function

File: gst/nnstreamer/tensor_source/tensor_src_iio.c [hunk 3]
```diff
@@ -873,6 +959,143 @@ gst_tensor_set_all_channels (GstTensorSrcIIO * self, const gint contents)
...
+    goto error_ret;
+  } else if (tensor_info_merged_size > NNS_TENSOR_SIZE_LIMIT) {
+    GST_ERROR_OBJECT (tensor_src_iio,
+        "Number of tensors required %u for data exceed the max limit",
+        tensor_info_merged_size);
+    goto error_ret;
+  }
+
+  /** tensor/s config data */
+  if (tensor_info_merged_size == 1) {
```
Comment (jaeyun-jung) [inline gst/nnstreamer/tensor_source/tensor_src_iio.c]: [trivial suggestion] use only tensors config (self->config)
1. remove GstTensorSrcIIODataConfig and set GstTensorsConfig in data structure
2. to get caps from config
if is_tensor
    GstTensorConfig c;
    gst_tensor_info_copy (&c.info, &self->config.info.info[0]);
    c.rate_n = self->config.rate_n;
    c.rate_d = self->config.rate_d;
    caps = gst_tensor_caps_from_config (&c);
else
    caps = gst_tensors_caps_from_config (&self->config);

Comment (kparichay) [reply gst/nnstreamer/tensor_source/tensor_src_iio.c]: This way would create an array for 16 (in other/tensors) but never use more than 1 element for other/tensor case. On the other hand, this method might make code simpler.
I am not sure which way is better - @myungjoo @again4you @leemgs - can you please comment on this?

File: gst/nnstreamer/tensor_source/tensor_src_iio.c (hunk 2 above)
Comment (kparichay) [reply gst/nnstreamer/tensor_source/tensor_src_iio.c]: Updated.
Also added graceful handling of failed merging.

File: gst/nnstreamer/tensor_source/tensor_src_iio.c [hunk 4]
```diff
@@ -322,8 +339,73 @@ gst_tensor_src_iio_init (GstTensorSrcIIO * self)
...
+  /** verify that all the types and shapes match */
+  for (info_idx = 0; info_idx < size; info_idx++) {
+    if (!gst_tensor_info_is_equal (info + base_idx, info + info_idx)) {
+      mismatch = TRUE;
+      break;
+    }
+  }
+
+  /** return original if cant be merged and size within limits */
+  if (mismatch || dim_avail) {
```
Comment (jaeyun-jung) [inline gst/nnstreamer/tensor_source/tensor_src_iio.c]: plz check the condition dim_avail

Comment (kparichay) [reply gst/nnstreamer/tensor_source/tensor_src_iio.c]: I missed it in the update.
Thanks for pointing out.
Done.

File: gst/nnstreamer/tensor_source/tensor_src_iio.c (hunk 3 above)
Comment (myungjoo) [reply gst/nnstreamer/tensor_source/tensor_src_iio.c]: You may neglect the wasted memory specifically for that; using TensorsConfig means that we've decided to neglect that waste for the sake of code generality. :)

As long as we use GstBuffer (it has a static array of 16), we always waste that amount of memory. It makes all codes to use the same data structure, making it easier to read and to maintain.

Comment (kparichay) [reply gst/nnstreamer/tensor_source/tensor_src_iio.c]: Updated, thanks for clarification.

**Comment by kparichay** [comment]
**V2**
- using only tensors config now
- removed bug of merging info of tensors

File: gst/nnstreamer/tensor_source/tensor_src_iio.c [hunk 5]
```diff
@@ -1008,15 +1229,83 @@ gst_tensor_src_iio_start (GstBaseSrc * src)
...
   /** complete the start of the base src */
   gst_base_src_start_complete (src, GST_FLOW_OK);
   return TRUE;
 
+error_caps_free:
+  gst_caps_unref (prev_caps);
+  gst_caps_unref (caps);
+
+error_config_free:
+  g_free (self->tensors_config);
```
Comment (jaeyun-jung) [inline gst/nnstreamer/tensor_source/tensor_src_iio.c]: In gst_tensor_src_iio_create_config()
```
gst_tensor_info_copy (&config->info.info[info_idx], &info[info_idx]);
...
tensor_src_iio->tensors_config = config;
```

before free tensors_config, it needs gst_tensors_info_free() (name newly allocated)

File: gst/nnstreamer/tensor_source/tensor_src_iio.c [hunk 6]
```diff
@@ -1047,6 +1336,8 @@ gst_tensor_src_iio_stop (GstBaseSrc * src)
 
   self->configured = FALSE;
 
+  g_free (self->tensors_config);
```
Comment (jaeyun-jung) [inline gst/nnstreamer/tensor_source/tensor_src_iio.c]: This line also needs free function for config-name)

Comment (kparichay) [reply gst/nnstreamer/tensor_source/tensor_src_iio.c]: Done.

File: gst/nnstreamer/tensor_source/tensor_src_iio.c (hunk 5 above)
Comment (kparichay) [reply gst/nnstreamer/tensor_source/tensor_src_iio.c]: Done.

**Comment by kparichay** [comment]
**V3**
Added freeing memory of `gchar *name` in `GstTensorInfo`

File: gst/nnstreamer/tensor_source/tensor_src_iio.c [hunk 7]
```diff
@@ -1008,15 +1230,86 @@ gst_tensor_src_iio_start (GstBaseSrc * src)
...
+
+  if (self->is_tensor) {
+    GstTensorConfig tensor_config;
+    gst_tensor_info_copy (&tensor_config.info,
+        &(self->tensors_config->info.info[0]));
+    tensor_config.rate_n = self->tensors_config->rate_n;
+    tensor_config.rate_d = self->tensors_config->rate_d;
+    caps = gst_tensor_caps_from_config (&tensor_config);
+    gst_tensor_info_free (&tensor_config.info);
+    g_free (&tensor_config);
```
Comment (jaeyun-jung) [inline gst/nnstreamer/tensor_source/tensor_src_iio.c]: remove this line

Comment (kparichay) [reply gst/nnstreamer/tensor_source/tensor_src_iio.c]: my mistake, done.

**Comment by kparichay** [comment]
**V3**
Added freeing memory of `gchar *name` in `GstTensorInfo`

**V4**
rebased

**Pull request merged**